src/tactic/norm_swap.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2021 Yakov Pechersky All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import data.equiv.basic
+import tactic.norm_num
+
+/-!
+# `norm_swap`
+
+Evaluating `swap x y z` for numerals `x y z : ℕ`.
+-/
+
+open equiv tactic expr
+
+/--
+Match a term of shape `app (⇑f) x` and return expressions for:
+  * the type `tf` of `f`
+  * the instance `has_coe_to_fun tf`
+  * `f`
+  * `x`
+  -/
+meta def expr.get_of_coe_fn (e : expr) (f : name) : tactic (expr × expr × expr × expr) :=
+do
+  if e.is_app_of ``coe_fn
+  then do
+    [α, inst, fexpr, x] ← pure e.get_app_args,
+    let fname : name := fexpr.get_app_fn.const_name,
+    if fname = f then pure (α, inst, fexpr, x)
+      else fail!"retrieved function name {fname} is not the expected {f}"
+  else fail "not of coe_fn form with a single argument"
+
+/--
+Creates the application `n c.α p l`, where `p` is a type class instance found in the cache `c`,
+but works on `c.α : Sort*` instead of how `mk_app` uses `c.α : Type*`.
+-/
+meta def tactic.instance_cache.mk_sorted_app (c : instance_cache) (n : name) (l : list expr) :
```

Comment:
The problem is in `mk_instance_cache`:
```
/-- Creates an `instance_cache` for the type `α`. -/
meta def mk_instance_cache (α : expr) : tactic instance_cache :=
do u ← mk_meta_univ,
   infer_type α >>= unify (expr.sort (level.succ u)),
   u ← get_univ_assignment u,
   return ⟨α, u, mk_name_map⟩
```
There, the stored `instance_cache.univ` is 1 less than what is needed in my case. That is, `mk_instance_cache` assumes that any use of the cache via `mk_app` will be with functions that expect a `Type*`. Since `coe_fn` is over `Sort*`, `mk_app` does not work for me.

@robertylewis would you suggest editing how the `instance_cache` works, or just avoiding it all together and making the raw `app` `expr`?

Reply:
I think the proper solution (which belongs in another PR) is to change `instance_cache.mk_app` to check whether the constant expects a `Type` or `Sort`. e.g.

```lean
open tactic 

meta def takes_type_arg (n : name) : tactic bool := 
do d ← get_decl n, 
match d.type with 
| expr.pi _ _ (expr.sort (level.succ _)) _ := return tt
| expr.pi _ _ (expr.sort _) _ := return ff
| _ := failed
end 

universe u
def f {α : Type u} := ℕ
def g {α : Sort u} := ℕ

#eval takes_type_arg `f >>= trace
#eval takes_type_arg `g >>= trace
```

Reply:
That said: you don't need to use a `instance_cache` for `coe_fn` at all in your code. The cache will never be populated with a `has_coe_to_fun` before you use it (`prove_ne` won't add this instance) and it's discarded right after. Just use regular `tactic.mk_app`.